### PR TITLE
Add flashlight mod

### PR DESCRIPTION
**Pull request by jorolf**
WIP: I still need the values/algorithms for the flashlight size

- [x] Depends on https://github.com/ppy/osu-resources/pull/45
- [x] Depends on https://github.com/ppy/osu-resources/pull/47

Known issues:
- ~~flashlight size doesn't take scale into account~~
---
Finally present in lazer

### CODE REVIEW

File: osu.Game.Rulesets.Osu/Mods/OsuModFlashlight.cs [hunk 1]
```diff
@@ -1,12 +1,96 @@
...
+
+        public void ApplyToRulesetContainer(RulesetContainer<OsuHitObject> rulesetContainer)
+        {
+            rulesetContainer.KeyBindingInputManager.Add(new Flashlight
+            {
+                RelativeSizeAxes = Axes.Both,
+            });
+        }
+
+        private class Flashlight : Drawable, IRequireHighFrequencyMousePosition
```
Comment (VPeruS) [inline osu.Game.Rulesets.Osu/Mods/OsuModFlashlight.cs]: Why not part of ModFlashlight? In that case you need just to set right settings for any mods(OsuFlashlight : Flashlight).

Comment (peppy) [reply osu.Game.Rulesets.Osu/Mods/OsuModFlashlight.cs]: If making that generic i’d call it `CircularFlashlightMask` or similar.

Comment (jorolf) [reply osu.Game.Rulesets.Osu/Mods/OsuModFlashlight.cs]: What about a box/circle boolean uniform inside the shader?

File: osu.Game.Rulesets.Osu/Mods/OsuModFlashlight.cs [hunk 2]
```diff
@@ -1,12 +1,96 @@
...
+            public Quad ScreenSpaceDrawQuad;
+            public MousePositionWrapper MousePosWrapper;
+
+            public override void Draw(Action<TexturedVertex2D> vertexAction)
+            {
+                base.Draw(vertexAction);
+
+                Shader.Bind();
+                // ReSharper disable once AssignmentInConditionalExpression
+                if(MousePosWrapper.FlashlightUniformUpdated = !MousePosWrapper.FlashlightUniformUpdated)
```
Comment (VPeruS) [inline osu.Game.Rulesets.Osu/Mods/OsuModFlashlight.cs]:
```suggestion
                MousePosWrapper.FlashlightUniformUpdated = !MousePosWrapper.FlashlightUniformUpdated
                if (MousePosWrapper.FlashlightUniformUpdated)
```

Comment (jorolf) [reply osu.Game.Rulesets.Osu/Mods/OsuModFlashlight.cs]: tbh that line doesn't even work because it gets inverted anyway instead of just firing once

Comment (VPeruS) [reply osu.Game.Rulesets.Osu/Mods/OsuModFlashlight.cs]: I thought you wanted to skip every second draw.

Comment (VPeruS) [reply osu.Game.Rulesets.Osu/Mods/OsuModFlashlight.cs]:
```suggestion
                
                if (!MousePosWrapper.FlashlightUniformUpdated) {}
                MousePosWrapper.FlashlightUniformUpdated = true
```

File: osu.Game/Rulesets/Mods/ModFlashlight.cs [hunk 3]
```diff
@@ -1,17 +1,158 @@
...
+
+        public class FlashlightUniformWrapper
+        {
+            public bool Rectangular;
+            public bool RectangularChanged = true;
+
+            public Vector2 FlashlightPosition;
+            public bool FlashlightPositionChanged = true;
+
+            public float CircularFlashlightSize;
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Mods/ModFlashlight.cs]: Like I mentioned in the other comment, I really don't think we need this separation of size, position, and the rectangular thing you have here. They should be re-usable for all flashlight implementations.

File: osu.Game/Rulesets/Mods/ModFlashlight.cs [hunk 4]
```diff
@@ -1,17 +1,158 @@
...
+            public bool CircularFlashlightSizeChanged = true;
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Mods/ModFlashlight.cs]: Is this some form of pre-optimisation? The shader already caches these values. Unnecessary.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Mania/Mods/ManiaModFlashlight.cs [hunk 5]
```diff
@@ -2,13 +2,56 @@
...
+
+            public override bool Invalidate(Invalidation invalidation = Invalidation.All, Drawable source = null, bool shallPropagate = true)
+            {
+                if ((invalidation & Invalidation.DrawSize) > 0)
+                {
+                    Schedule(() =>
+                    {
+                        FlashlightSize = new Vector2(DrawWidth, FlashlightSize.Y);
+
+                        FlashlightPosition = DrawPosition + DrawSize / 2;
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Mania/Mods/ManiaModFlashlight.cs]: It's easier to set these values every frame in `Update()`, the base `FlashlightSize`/`FlashlightPosition` setters handle de-duping of values.

File: osu.Game.Rulesets.Catch/Mods/CatchModFlashlight.cs [hunk 6]
```diff
@@ -1,12 +1,66 @@
...
+                FlashlightSize = new Vector2(0, getSizeFor(0));
+            }
+
+            protected override void Update()
+            {
+                base.Update();
+
+                var catcher = playfield.CatcherArea.MovableCatcher;
+
+                FlashlightPosition = catcher.ToSpaceOfOtherDrawable(catcher.Position, this);
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Catch/Mods/CatchModFlashlight.cs]: This should be more like:

```
var catcherArea = playfield.CatcherArea;

FlashlightPosition = catcherArea.ToSpaceOfOtherDrawable(catcherArea.MovableCatcher.DrawPosition, this);
```

As it is, this causes the flashlight to move as the catcher's direction changes.

File: osu.Game.Rulesets.Taiko/Mods/TaikoModFlashlight.cs [hunk 7]
```diff
@@ -1,12 +1,77 @@
...
+                this.TransformTo(nameof(FlashlightSize), new Vector2(0, getSizeFor(newCombo)), FLASHLIGHT_FADE_DURATION);
+            }
+
+            protected override string FragmentShader => "CircularFlashlight";
+
+            public override bool Invalidate(Invalidation invalidation = Invalidation.All, Drawable source = null, bool shallPropagate = true)
+            {
+                if ((invalidation & Invalidation.DrawSize) > 0)
+                {
+                    Schedule(() =>
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Taiko/Mods/TaikoModFlashlight.cs]: This is kind-of a new pattern that we haven't used elsewhere, and may lead to issues such as multiple invalidations per-frame causing multiple schedules.

Better than this is to use a `Cached` value which you validate inside `Update()` - check how it's done in `SpriteText` / `FlowContainer`, etc for reference.

File: osu.Game/Rulesets/Mods/ModFlashlight.cs [hunk 8]
```diff
@@ -1,17 +1,158 @@
...
+                {
+                    if (flashlightSize == value) return;
+
+                    flashlightSize = value;
+                    Invalidate(Invalidation.DrawNode);
+                }
+            }
+        }
+
+        public struct FlashlightUniformWrapper
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Mods/ModFlashlight.cs]: Still necessary? Might as well merge it into the drawnode tbh, it's a bit awkward that it's called "MousePosWrapper", and "UniformWrapper" when it's really neither of those.

**Review by smoogipoo — CHANGES_REQUESTED** [review]
It's getting there, think it should be fine after these.

**Comment by smoogipoo** [comment]
Will need rebalancing later, but this is a good start 👍

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**